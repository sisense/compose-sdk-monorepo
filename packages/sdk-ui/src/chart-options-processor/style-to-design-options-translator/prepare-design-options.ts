import {
  CartesianChartDataOptionsInternal,
  ChartDataOptionsInternal,
  SeriesStyleOptions,
  Value,
} from '@/chart-data-options/types';
import { ChartStyleOptions, ChartType } from '@/types';
import { WithRequiredProp } from '@/utils/utility-types';
import pick from 'lodash-es/pick';
import merge from 'ts-deepmerge';
import { getDefaultStyleOptions } from '../chart-options-service';
import { DesignOptions, isCartesian, SeriesDesignOptions } from '../translations/types';
import { translateStyleOptionsToDesignOptions } from './translate-style-to-design-options';
import { getSeriesChartDesignOptions } from './translate-to-highcharts-options';

export function prepareChartDesignOptions(
  chartType: ChartType,
  dataOptionsInternal: ChartDataOptionsInternal,
  styleOptions?: ChartStyleOptions,
): DesignOptions {
  const styleOptionsWithDefaults = extendStyleOptionsWithDefaults(
    styleOptions ?? {},
    getDefaultStyleOptions(),
  );
  return translateStyleOptionsToDesignOptions(
    chartType,
    styleOptionsWithDefaults,
    dataOptionsInternal,
  );
}

export function extendStyleOptionsWithDefaults(
  styleOptions: ChartStyleOptions,
  defaults: ChartStyleOptions,
): ChartStyleOptions {
  return merge.withOptions(
    {
      mergeArrays: false,
    },
    defaults,
    styleOptions,
  ) as ChartStyleOptions;
}

export function getDesignOptionsPerSeries(
  dataOptionsInternal: ChartDataOptionsInternal,
  chartType: ChartType,
  styleOptions: ChartStyleOptions,
): Record<string, SeriesDesignOptions> {
  const seriesDesignOptions = getAllSeriesStyleOptions(dataOptionsInternal, chartType).map(
    (styleOptionsForSpecificSeries) => {
      const seriesId = styleOptionsForSpecificSeries.seriesId;
      const seriesStyleOptions = styleOptionsForSpecificSeries.seriesStyleOptions;
      const completeDesignOptions = getSeriesChartDesignOptions(
        chartType,
        extendStyleOptionsWithDefaults(seriesStyleOptions, styleOptions),
      );
      return {
        seriesId,
        designOptions: pick(completeDesignOptions, ['lineWidth', 'marker']),
      };
    },
  );
  return seriesDesignOptions.reduce(
    (acc, { seriesId, designOptions }) => ({ ...acc, [seriesId]: designOptions }),
    {},
  );
}

function getAllSeriesStyleOptions(
  dataOptions: ChartDataOptionsInternal,
  chartType: ChartType,
): StyleOptionsForSpecificSeries[] {
  if (isCartesian(chartType)) {
    return getStyleOptionsPerSeriesFromCartesianDataOptions(
      dataOptions as CartesianChartDataOptionsInternal,
    );
  }
  return [];
}

type StyleOptionsForSpecificSeries = {
  seriesId: string;
  seriesStyleOptions: SeriesStyleOptions;
};

function getStyleOptionsPerSeriesFromCartesianDataOptions(
  dataOptions: CartesianChartDataOptionsInternal,
): StyleOptionsForSpecificSeries[] {
  return dataOptions.y.filter(hasDefinedStyleOptionsForSeries).map((y) => ({
    seriesId: y.name,
    seriesStyleOptions: y.seriesStyleOptions,
  }));
}

function hasDefinedStyleOptionsForSeries(
  value: Value,
): value is WithRequiredProp<Value, 'seriesStyleOptions'> {
  return value.seriesStyleOptions !== undefined;
}
