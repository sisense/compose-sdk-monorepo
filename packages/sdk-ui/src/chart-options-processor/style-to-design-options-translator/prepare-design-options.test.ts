/** @vitest-environment jsdom */
import {
  AreamapChartDataOptionsInternal,
  CartesianChartDataOptionsInternal,
  CategoricalChartDataOptionsInternal,
  ScattermapChartDataOptionsInternal,
} from '@/chart-data-options/types';
import {
  AreamapStyleOptions,
  FunnelStyleOptions,
  LineStyleOptions,
  ScattermapStyleOptions,
} from '@/types';
import { prepareChartDesignOptions } from './prepare-design-options';

describe('prepareChartDesignOptions', () => {
  it('should return design options for combo "line + area" chart', () => {
    const chartType = 'line';
    const dataOptionsInternal: CartesianChartDataOptionsInternal = {
      x: [
        {
          column: {
            name: 'Months',
            type: 'datelevel',
          },
          dateFormat: 'yy-MM',
        },
      ],
      y: [
        { column: { name: '$measure0_avg Cost', title: 'avg Cost' } },
        {
          column: {
            name: '$measure1_Total Revenue',
            title: 'Total Revenue',
          },
          seriesStyleOptions: {
            lineWidth: {
              width: 'thick',
            },
            markers: {
              enabled: true,
              fill: 'filled',
              size: 'large',
            },
          },
        },
      ],
      breakBy: [],
    };
    const styleOptions: LineStyleOptions = {
      subtype: 'line/basic',
      lineWidth: {
        width: 'thin',
      },
      yAxis: {
        title: {
          enabled: true,
          text: 'AVG COST',
        },
      },
      y2Axis: {
        title: {
          enabled: true,
          text: 'REVENUE',
        },
      },
      markers: {
        enabled: true,
        fill: 'hollow',
        size: 'small',
      },
      seriesLabels: {
        enabled: true,
        rotation: -90,
      },
    };
    const result = prepareChartDesignOptions(chartType, dataOptionsInternal, styleOptions);
    expect(result).toMatchSnapshot();
  });

  it('should return design options for areamap chart', () => {
    const chartType = 'areamap';
    const dataOptionsInternal: AreamapChartDataOptionsInternal = {
      geo: {
        column: {
          name: 'Country',
          type: 'text-attribute',
        },
      },
      color: {
        column: {
          name: '$measure0_sum Cost',
          title: 'Total Cost',
        },
      },
    };
    const styleOptions: AreamapStyleOptions = {
      mapType: 'world',
    };
    const result = prepareChartDesignOptions(chartType, dataOptionsInternal, styleOptions);
    expect(result).toMatchSnapshot();
  });

  it('should return design options for scattermap chart', () => {
    const chartType = 'scattermap';
    const dataOptionsInternal: ScattermapChartDataOptionsInternal = {
      locations: [
        {
          column: {
            name: 'Country',
            type: 'text-attribute',
          },
        },
      ],
      locationLevel: 'auto',
    };
    const styleOptions: ScattermapStyleOptions = {
      markers: {
        fill: 'hollow-bold',
      },
    };
    const result = prepareChartDesignOptions(chartType, dataOptionsInternal, styleOptions);
    expect(result).toMatchSnapshot();
  });

  it('should return design options for funnel chart', () => {
    const chartType = 'funnel';
    const dataOptionsInternal: CategoricalChartDataOptionsInternal = {
      y: [
        {
          column: {
            title: 'Unique Users',
            aggregation: 'sum',
            name: '$measure0_Unique Users',
          },
          enabled: true,
          sortType: 'sortDesc',
        },
      ],
      breakBy: [{ column: { name: 'Stage', type: 'string' } }],
    };
    const styleOptions: FunnelStyleOptions = {
      legend: {
        enabled: true,
        position: 'left',
      },
      labels: {
        categories: true,
        decimals: true,
        enabled: true,
        percent: true,
        value: true,
      },
      funnelDirection: 'regular',
      dataLimits: {
        seriesCapacity: 3,
      },
    };
    const result = prepareChartDesignOptions(chartType, dataOptionsInternal, styleOptions);
    expect(result).toMatchSnapshot();
  });
});
