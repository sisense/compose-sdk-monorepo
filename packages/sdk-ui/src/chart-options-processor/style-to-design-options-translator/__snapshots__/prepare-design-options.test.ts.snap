// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`prepareChartDesignOptions > should return design options for areamap chart 1`] = `
{
  "autoZoom": {
    "enabled": true,
  },
  "dataLimits": {
    "categoriesCapacity": 100000,
    "seriesCapacity": 50,
  },
  "funnelDirection": "regular",
  "funnelLabels": {
    "enabled": true,
    "showCategories": true,
    "showDecimals": false,
    "showPercent": true,
    "showValue": true,
  },
  "funnelSize": "regular",
  "funnelType": "regular",
  "indicatorComponents": {
    "labels": {
      "shouldBeShown": true,
    },
    "secondaryTitle": {
      "text": "",
    },
    "ticks": {
      "shouldBeShown": true,
    },
    "title": {
      "shouldBeShown": false,
      "text": "",
    },
  },
  "indicatorType": "numeric",
  "legend": {
    "enabled": true,
    "position": "bottom",
  },
  "lineType": "straight",
  "lineWidth": 2,
  "mapType": "world",
  "marker": {
    "enabled": false,
    "fill": "full",
    "size": "small",
  },
  "numericSubtype": "numericSimple",
  "pieType": "classic",
  "skin": "vertical",
  "stackType": "classic",
  "xAxis": {
    "enabled": true,
    "gridLine": true,
    "labels": true,
    "max": null,
    "min": null,
    "tickInterval": null,
    "title": "X Axis title",
    "titleEnabled": true,
    "type": "linear",
  },
  "yAxis": {
    "enabled": true,
    "gridLine": true,
    "labels": true,
    "max": null,
    "min": null,
    "tickInterval": null,
    "title": "Y Axis title",
    "titleEnabled": true,
    "type": "linear",
  },
}
`;

exports[`prepareChartDesignOptions > should return design options for combo "line + area" chart 1`] = `
{
  "autoZoom": {
    "enabled": true,
    "scrollerLocation": undefined,
  },
  "dataLimits": {
    "categoriesCapacity": 100000,
    "seriesCapacity": 50,
  },
  "designPerSeries": {
    "$measure1_Total Revenue": {
      "marker": {
        "enabled": true,
        "fill": "full",
        "size": "large",
      },
    },
  },
  "funnelDirection": "regular",
  "funnelLabels": {
    "enabled": true,
    "showCategories": true,
    "showDecimals": false,
    "showPercent": true,
    "showValue": true,
  },
  "funnelSize": "regular",
  "funnelType": "regular",
  "indicatorComponents": {
    "labels": {
      "shouldBeShown": true,
    },
    "secondaryTitle": {
      "text": "",
    },
    "ticks": {
      "shouldBeShown": true,
    },
    "title": {
      "shouldBeShown": false,
      "text": "",
    },
  },
  "indicatorType": "numeric",
  "legend": {
    "enabled": true,
    "position": "bottom",
  },
  "line": {
    "dashStyle": undefined,
    "endCap": undefined,
    "shadow": undefined,
    "width": 1,
  },
  "lineType": "straight",
  "lineWidth": 2,
  "marker": {
    "enabled": true,
    "fill": "hollow",
    "size": "small",
  },
  "numericSubtype": "numericSimple",
  "pieType": "classic",
  "seriesLabels": {
    "enabled": true,
    "rotation": -90,
  },
  "skin": "vertical",
  "stackType": "classic",
  "xAxis": {
    "enabled": true,
    "gridLine": false,
    "labels": true,
    "max": null,
    "min": null,
    "tickInterval": null,
    "title": "X Axis title",
    "titleEnabled": false,
    "type": "linear",
  },
  "yAxis": {
    "enabled": true,
    "gridLine": true,
    "labels": true,
    "max": null,
    "min": null,
    "tickInterval": null,
    "title": "AVG COST",
    "titleEnabled": true,
    "type": "linear",
  },
}
`;

exports[`prepareChartDesignOptions > should return design options for funnel chart 1`] = `
{
  "autoZoom": {
    "enabled": true,
  },
  "dataLimits": {
    "categoriesCapacity": 50000,
    "seriesCapacity": 3,
  },
  "funnelDirection": "regular",
  "funnelLabels": {
    "enabled": true,
    "showCategories": true,
    "showDecimals": true,
    "showPercent": true,
    "showValue": true,
  },
  "funnelSize": "regular",
  "funnelType": "regular",
  "indicatorComponents": {
    "labels": {
      "shouldBeShown": true,
    },
    "secondaryTitle": {
      "text": "",
    },
    "ticks": {
      "shouldBeShown": true,
    },
    "title": {
      "shouldBeShown": false,
      "text": "",
    },
  },
  "indicatorType": "numeric",
  "legend": {
    "enabled": true,
    "position": "left",
  },
  "lineType": "straight",
  "lineWidth": 2,
  "marker": {
    "enabled": false,
    "fill": "full",
    "size": "small",
  },
  "numericSubtype": "numericSimple",
  "pieType": "classic",
  "skin": "vertical",
  "stackType": "classic",
  "xAxis": {
    "enabled": true,
    "gridLine": true,
    "labels": true,
    "max": null,
    "min": null,
    "tickInterval": null,
    "title": "X Axis title",
    "titleEnabled": true,
    "type": "linear",
  },
  "yAxis": {
    "enabled": true,
    "gridLine": true,
    "labels": true,
    "max": null,
    "min": null,
    "tickInterval": null,
    "title": "Y Axis title",
    "titleEnabled": true,
    "type": "linear",
  },
}
`;

exports[`prepareChartDesignOptions > should return design options for scattermap chart 1`] = `
{
  "autoZoom": {
    "enabled": true,
  },
  "dataLimits": {
    "categoriesCapacity": 100000,
    "seriesCapacity": 50,
  },
  "funnelDirection": "regular",
  "funnelLabels": {
    "enabled": true,
    "showCategories": true,
    "showDecimals": false,
    "showPercent": true,
    "showValue": true,
  },
  "funnelSize": "regular",
  "funnelType": "regular",
  "indicatorComponents": {
    "labels": {
      "shouldBeShown": true,
    },
    "secondaryTitle": {
      "text": "",
    },
    "ticks": {
      "shouldBeShown": true,
    },
    "title": {
      "shouldBeShown": false,
      "text": "",
    },
  },
  "indicatorType": "numeric",
  "legend": {
    "enabled": true,
    "position": "bottom",
  },
  "lineType": "straight",
  "lineWidth": 2,
  "marker": {
    "enabled": false,
    "fill": "full",
    "size": "small",
  },
  "markers": {
    "enabled": false,
    "fill": "hollow-bold",
    "size": "small",
  },
  "numericSubtype": "numericSimple",
  "pieType": "classic",
  "skin": "vertical",
  "stackType": "classic",
  "xAxis": {
    "enabled": true,
    "gridLine": true,
    "labels": true,
    "max": null,
    "min": null,
    "tickInterval": null,
    "title": "X Axis title",
    "titleEnabled": true,
    "type": "linear",
  },
  "yAxis": {
    "enabled": true,
    "gridLine": true,
    "labels": true,
    "max": null,
    "min": null,
    "tickInterval": null,
    "title": "Y Axis title",
    "titleEnabled": true,
    "type": "linear",
  },
}
`;

exports[`prepareChartDesignOptions > should throw an error if internal data options is not compatible with chart type (for restructured charts) 1`] = `[Error: Data options are not correctly converted for 'column' chart]`;

exports[`prepareChartDesignOptions > should throw an error if style options is not compatible with chart type (for restructured charts) 1`] = `[Error: Invalid style options for 'column' chart]`;
